sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -319,149 +328,193 @@ public static Read read() {
-      abstract Builder setFlattenResults(Boolean flattenResults);
-      abstract Builder setUseLegacySql(Boolean useLegacySql);
+    public static final TableRowParserFactory INSTANCE = new TableRowParserFactory();
 
-      abstract Builder setWithTemplateCompatibility(Boolean useTemplateCompatibility);
+    public SerializableFunction<GenericRecord, TableRow> apply(TableSchema tableSchema) {
+      final String jsonSchema;
+      try {
+        jsonSchema = BigQueryIO.JSON_FACTORY.toString(tableSchema);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
 
-      abstract Builder setBigQueryServices(BigQueryServices bigQueryServices);
-      abstract Read build();
-    }
+      return new SerializableFunction<GenericRecord, TableRow>() {
+        private Supplier<TableSchema> schema = Suppliers.memoize(
+            Suppliers.compose(new TableSchemaFunction(), Suppliers.ofInstance(jsonSchema)));
 
-    /** Ensures that methods of the from() / fromQuery() family are called at most once. */
-    private void ensureFromNotCalledYet() {
-      checkState(
-          getJsonTableRef() == null && getQuery() == null, "from() or fromQuery() already called");
+        @Override
+        public TableRow apply(GenericRecord input) {
+          return BigQueryAvroUtils.convertGenericRecordToTableRow(input, schema.get());
+        }
+      };
     }
+  }
 
-    /**
-     * Reads a BigQuery table specified as {@code "[project_id]:[dataset_id].[table_id]"} or
-     * {@code "[dataset_id].[table_id]"} for tables within the current project.
-     */
-    public Read from(String tableSpec) {
-      return from(StaticValueProvider.of(tableSpec));
-    }
+  /** Implementation of {@link #read}. */
+  @AutoValue
+  public abstract static class Read extends ReadBase<TableRow, Read, Read.Builder> {
 
-    /** Same as {@code from(String)}, but with a {@link ValueProvider}. */
-    public Read from(ValueProvider<String> tableSpec) {
-      ensureFromNotCalledYet();
-      return toBuilder()
-          .setJsonTableRef(
-              NestedValueProvider.of(
-                  NestedValueProvider.of(tableSpec, new TableSpecToTableRef()),
-                  new TableRefToJson()))
-          .build();
-    }
+    abstract Builder toBuilder();
 
-    /**
-     * Reads results received after executing the given query.
-     *
-     * <p>By default, the query results will be flattened -- see "flattenResults" in the <a
-     * href="https://cloud.google.com/bigquery/docs/reference/v2/jobs">Jobs documentation</a> for
-     * more information. To disable flattening, use {@link BigQueryIO.Read#withoutResultFlattening}.
-     *
-     * <p>By default, the query will use BigQuery's legacy SQL dialect. To use the BigQuery Standard
-     * SQL dialect, use {@link BigQueryIO.Read#usingStandardSql}.
-     */
-    public Read fromQuery(String query) {
-      return fromQuery(StaticValueProvider.of(query));
+    @Override
+    Coder<TableRow> getCoder() {
```

Comment:
I would prefer to use regular setters in the read() factory method, rather than method overrides, to set property values.

Reply:
👍 makes sense.

Reply:
ah upon trying to refactor this I realized why these are like they are: AutoValue has issues generating the code for `Read` since the parameter is "reified" into TableRow.

```
[ERROR] /Users/sniemitz/dev/beam/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java:[395,24] Parameter type org.apache.beam.sdk.coders.Coder<com.google.api.services.bigquery.model.TableRow> of setter method should be org.apache.beam.sdk.coders.Coder<T> to match getter org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.Read.getCoder
```

Reply:
Hmm, I think it may be better to just bite the bullet and have some boilerplate (separate builder classes for Read and ReadGenericRecords, perhaps the former delegating to the latter like in TextIO.Write): if today it required hackery to get to compile, one day it might require hackery for somebody to fix it after an innocuous change that will somehow be "too much" for AutoValue.

Reply:
Yeah, at this point it probably makes sense to just ditch the AutoValue stuff here, I don't know if its adding much value (lol no pun intended).

I'll give that a try.

Reply:
Based on the suggestion below of `Read` simply wrapping `ReadGenericRecords`, this becomes much simpler since I no longer need a base class for the PTransform and Builders.